### DISCUSSION TITLE
Clarification on code for preventing second preimage attacks in Merkle trees

### DISCUSSION THREAD

**Discussion by uddercover** [Cyfrin/foundry-full-course-cu#2503] [Q&A] [answered]
In the "Airdrop and signatures" section of the course, the lady uses this code to create the leaf: <code>bytes32 leaf = keccak256(bytes.concat(keccak256(abi.encode(account, amount))));</code>
What does <code>bytes.concat</code> do in the above line? Is it really necessary?

**Comment by EngrPips** [comment]
Hello @uddercover, can you try to do the process without `bytes.concat` and see if you will get the same output as the lady.

**Reply by uddercover** [reply, marked as answer]
I figured it out. 
The hash function <code>keccak256</code> takes a single bytes argument and returns a bytes32. As a result, the argument of the second <code>keccak256</code> would be a bytes32 instead of a bytes. So it has to be converted to a bytes type, but doing so explicitly is not allowed. The <code>bytes.concat</code> does this as it takes any bytes argument (bytes32 or otherwise) and returns it as a bytes type. I was curious because at face value, <code>bytes.concat</code> is used for concatenating variables, so it felt very random to just have it in there with only one argument.

**Reply by EngrPips** [reply]
Interesting.
